app/javascript/components/hocs/with-widget-transition-controls.jsx

```diff
@@ -0,0 +1,144 @@
+import React, { useEffect, useReducer } from 'react';
+import PropTypes from 'prop-types';
+import { mathMod } from 'ramda';
+
+const prefetchDelay = 1500;
+
+const shouldRequestWidget = (state, msTimeElapsedSinceStart) =>
+  state.msDurationRemaining < msTimeElapsedSinceStart;
+
+const shouldRequestPrefetch = (state, msTimeElapsedSinceStart) =>
+  state.msDurationRemaining - msTimeElapsedSinceStart < prefetchDelay;
+
+const isPaused = state => state.startTimestamp === null;
+
+const reducer = (state, action) => {
+  const { startTimestamp, msDurationRemaining } = state;
+  const now = Date.now();
+
+  switch (action.type) {
+    case 'togglePause':
+      return {
+        msDurationRemaining: isPaused(state)
+          ? msDurationRemaining
+          : msDurationRemaining - (now - startTimestamp),
+        startTimestamp: isPaused(state) ? now : null,
+      };
+    case 'startTimer':
+      return {
+        msDurationRemaining: action.msDurationSeconds,
+        startTimestamp: isPaused(state) ? null : now,
+      };
+    default:
+      return state;
+  }
+};
+
+/*
+ * WithWidgetControls returns a component which expects props that match
+ * WrappedComponent's props
+ */
+const WithWidgetTransitionControls = ({
+  WrappedComponent,
+  requestPrefetch,
+  requestWidget,
+  cityName,
+  widgets,
+}) => {
+  const { byIdOrFirst: current, enabled: enabledWidgets } = widgets;
+
+  const widgetTransitionTime = current.durationSeconds * 1000;
+
+  const index = enabledWidgets.findIndex(w => w.id === current.id);
+
+  const prevWidgetId =
+    enabledWidgets[mathMod(index - 1, enabledWidgets.length)].id;
+  const nextWidgetId =
+    enabledWidgets[mathMod(index + 1, enabledWidgets.length)].id;
+
+  const [state, dispatch] = useReducer(reducer, {
+    startTimestamp: Date.now(),
+    msDurationRemaining: null,
+  });
+
+  useEffect(() => {
+    dispatch({
+      type: 'startTimer',
+      msDurationSeconds: widgetTransitionTime,
+    });
+  }, [current.id]);
+
+  useEffect(() => {
+    const timerId = setInterval(() => {
+      const { startTimestamp } = state;
+      if (isPaused(state)) {
+        return;
+      }
+
+      const now = Date.now();
+      const timeElapsedSinceStart = now - startTimestamp;
+
+      if (shouldRequestWidget(state, timeElapsedSinceStart)) {
```

Comment:
Nit: `timeElapsedSinceStart` is a calculation from now and state. `shouldRequestWidget` / `shouldRequestPrefecth` are like selectors so I'd argue we should be passing in `now` and letting it do the `timeElapsedSinceStart` calculation from state. You could even have another selector just for that calculation.